File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

pair<int,int> diameterBT_app2(TreeNode* root){
    // base case :
    if(root == NULL){
        pair<int,int> p = make_pair(0,0);
        return p;
    }

    // recursive step :
    pair<int,int> left = diameterBT_app2(root->left);
    pair<int,int> right = diameterBT_app2(root->right);

    int op1 = left.first;
    int op2 = right.first;
    int op3 = left.second + right.second ;

    pair<int,int> ans;
    ans.first = max(op1,max(op2,op3));
    ans.second = max(left.second,right.second) + 1;

    return ans;

}
    int diameterOfBinaryTree(TreeNode* root) {
        return diameterBT_app2(root).first;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int d;

    
    public int diameterOfBinaryTree(TreeNode root) {
        d=0;
        diameter(root);
        return d;
    }

    private int diameter(TreeNode root){
        if(root==null)return 0;

        int l = diameter(root.left);
        int r = diameter(root.right);
        d = Math.max(d,l+r);
        return Math.max(l,r)+1;
    }
}